### PR TITLE
SI-7598 Make erasure of isInstanceOf[X.type] $outer-aware

### PR DESCRIPTION
**Pull request by retronym**
Type tests against Singleton-, This- and, Super- types are translated
into reference equality checks in erasure.

For instance:

```
class C { object X; X.isInstanceOf[X.type] }
```

Results in:

```
class C { object X; X eq X }
```

But, the term tree created to refer to the object behind the
SingleType was constructed without being explicit about outer
pointers.

After typer:

```
class C { object X; class Inner { X.isInstanceOf[X.type] } }
```

After explicitouter:

```
class C { object X; class Inner($outer: C) { $outer().X().isInstanceOf[X.type] } }
```

After erasure:

```
class C { object X; class Inner($outer: C) { $outer().X() eq Outer.this.X() } }
```

That leads to a crash in icode.

This commit changes the explicit outer tree transformer to transform
the types of TypeTrees:

After exlicitouter:

```
Inner.this.$outer().X().isInstanceOf[Inner.this.$outer.X.type](
```

Do to this, we need versions of outerValue, outerSelect, and outerPath
that operate on types. I also refactored the term-level versions of
these to help emphasise the similarities and differences with the new
versions.

Some robustness is needed in the face of ill-scoped references left
in the trees after extension methods and tail call elimination have
left a mess in pos/t6891.scala.

This is a resubmission of #3066 that addresses the problem at the source
(explicitouter) rather than in erasure.

Review by @gkossakowski / @odersky .

### CODE REVIEW

**Comment by retronym** [comment]
Looks like I might have an infinite loop in there, judging from:

```
[INFO] /localhome/jenkins/b/workspace/pr-scala-integrate-ide/target/scala-refactoring/org.scala-refactoring.library/src/test/java:-1: info: compiling
[INFO] Compiling 115 source files to /localhome/jenkins/b/workspace/pr-scala-integrate-ide/target/scala-refactoring/org.scala-refactoring.library/target/classes at 1382619280843
[WARNING] /localhome/jenkins/b/workspace/pr-scala-integrate-ide/target/scala-refactoring/org.scala-refactoring.library/src/main/scala/scala/tools/refactoring/transformation/TreeFactory.scala:125: warning: a pure expression does nothing in statement position; you may be omitting necessary parentheses
[WARNING]         case t => t
[WARNING]                   ^
Build timed out (after 240 minutes). Marking the build as aborted.
Build was aborted
Archiving artifacts
```

I would still welcome a comparative review of this PR vs #3066

**Comment by gkossakowski** [comment]
@retronym: I'll have a look tomorrow.

**Comment by JamesIry** [comment]
ping @gkossakowski

File: src/compiler/scala/tools/nsc/transform/ExplicitOuter.scala [hunk 1]
```diff
@@ -225,10 +225,8 @@ abstract class ExplicitOuter extends InfoTransform
      *
      * Will return `EmptyTree` if there is no outer accessor because of a premature self reference.
      */
-    protected def outerValue: Tree = outerParam match {
-      case NoSymbol   => outerSelect(gen.mkAttributedThis(currentClass))
-      case outerParam => gen.mkAttributedIdent(outerParam)
-    }
+    protected def outerValue    : Tree = outerParam.fold(outerSelect(gen.mkAttributedThis(currentClass)))(gen.mkAttributedIdent)
```
Comment (gkossakowski) [inline src/compiler/scala/tools/nsc/transform/ExplicitOuter.scala:228]: I'm not exactly sure if this change improves readability.

File: src/compiler/scala/tools/nsc/transform/ExplicitOuter.scala [hunk 2]
```diff
@@ -240,28 +238,40 @@ abstract class ExplicitOuter extends InfoTransform
...
+      else singleType(base, outerAcc)
+    }
+
+    /** Construct one of:
+      * {{{
+      *   $base.`$outer `   // outer field selection
+      *   $base.`$outer`()  // outer accessor selection
+      * }}}
+      *
+      * The former is only possible if the
```
Comment (gkossakowski) [inline src/compiler/scala/tools/nsc/transform/ExplicitOuter.scala:260]: Interrupted comment.

File: src/compiler/scala/tools/nsc/transform/ExplicitOuter.scala (hunk 1 above)
Comment (retronym) [inline src/compiler/scala/tools/nsc/transform/ExplicitOuter.scala:228]: the idea is to show the to variations of this (term/type) side-by-side.

Comment (gkossakowski) [inline src/compiler/scala/tools/nsc/transform/ExplicitOuter.scala:228]: Yes, I noticed that. I'm not a fan of fold on symbol/boolean/option but in this case it's ok I guess.

**Comment by gkossakowski** [comment]
PLS REBUILD ALL

**Comment by gkossakowski** [comment]
This looks good to me: nice refactoring and fix. Let's see if tests pass or they go into infinite loop again.

**Comment by retronym** [comment]
I assume the failure was legit the first time; it needs investigation. I saw similar problems when developing the patch.

**Pull request closed without merging**